pkg/mover/intermediate.go

```diff
@@ -0,0 +1,106 @@
+// Copyright 2021 VMware, Inc.
+// SPDX-License-Identifier: BSD-2-Clause
+
+package mover
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/google/go-containerregistry/pkg/name"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/tarball"
+	"github.com/vmware-tanzu/asset-relocation-tool-for-kubernetes/internal"
+	"helm.sh/helm/v3/pkg/chart"
+)
+
+var (
+	// ErrNotIntermediateBundle when a verified path does not have expected intermediate bundle contents
+	ErrNotIntermediateBundle = errors.New("not an intermediate chart bundle")
+)
+
+func saveIntermediateBundle(cd *ChartData, tarFile string, log Logger) error {
+	tfw, err := newTarFileWriter(tarFile)
+	if err != nil {
+		return err
+	}
+
+	// hints file goes first to be extracted quickly on demand
+	log.Printf("Writing %s...\n", HintsFilename)
+	if err := tfw.WriteFile(HintsFilename, cd.rawHints, os.ModePerm); err != nil {
+		return fmt.Errorf("failed to write %s: %w", HintsFilename, err)
+	}
+
+	log.Printf("Writing Helm Chart files at %s/...\n", cd.chart.Metadata.Name)
+	if err := tarChart(tfw, cd.chart); err != nil {
+		return fmt.Errorf("failed archiving original-chart/: %w", err)
+	}
+	if err := tfw.Close(); err != nil {
```

Comment:
OK,

Why do I close it... this is kind of tricky. The image files are to be appended after the tar is already created and has the hints and chart files. If you do this (done in the RawWriter method):
~~~go
	// this flush here allows for another tar writer to continue on the stream
	tfw.Flush()
	return tfw.WriteCloser
~~~

So in theory closing is not needed. But it felt weird, because the idea is that the tarball lib will use its own tar writer to append stuff there, but then you must ensure the stream is ready for that.

I am really not sure how to make this less hacky and more clear.

Reply:
I think I found a better way.